31823 Improved Logging Information in List Table

johnregan3 opened the pull request:
Closes #31823 .

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

* Adds "View Details"  button to toggle display of detailed Log information.

![Screen Shot 2022-04-21 at 9 59 16 AM](https://user-images.githubusercontent.com/2053940/164490092-cc1711d8-4978-43c7-aa5a-07f399cc5688.png)

![Screen Shot 2022-04-21 at 9 59 29 AM](https://user-images.githubusercontent.com/2053940/164490123-5a9564d2-0d88-48bf-8e7b-a3c077037cf0.png)


### How to test the changes in this Pull Request:

1. Create an error. In (plugins/woocommerce/includes/abstracts/abstract-wc-order.php) add this code inside the `try` block (around line 178):
```
throw new ErrorException( 'test message', 0, 1, __FILE__, 1 );
```
2. Go into any Order in the WP Admin, then click the "Update" button.
3. Remove the above testing code.
4. Visit Status > Logs (wp-admin/admin.php?page=wc-status&tab=logs).
5. Click the "View Details" button in the "Details" column.

### Other information:

Note that with the new precommit hook, I had to `ignore` some specific PHPCS errors outside the scope of this task.

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

### Changelog entry

> Add improved display of Error Log details.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

johnregan3 committed:
31823 WIP  Initial work

johnregan3 committed:
Merge branch 'trunk' into 31823-improved-logging-access

johnregan3 committed:
Merge remote-tracking branch 'upstream/trunk' into 31823-improved-logging-access

johnregan3 committed:
Remove quick edit filters

johnregan3 committed:
31823 | Error Details Toggles & PHPCS

johnregan3 committed:
31823 - Two col error detail display

johnregan3 committed:
31823 - Handle any context
not only orders and errors.
Also adds readonly to the textarea(s)

johnregan3 commented:
@barryhughes Updated this PR to not use hardcoded Context information (that is, "order" and "error").  Also adds `readonly` to the textarea to prevent users from thinking they can modify the data.

barryhughes commented:
Sounds good. Just to set expectations, it may take another week or so before I have space to review (but it's on the list).

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -255,6 +274,82 @@ protected function source_dropdown() {
+	public function render_log( $key, $log ) {
+		// This is unserializing "safe" data from the database, not user input.
+		$context = unserialize( $log['context'] );
+		$context = $this->to_multidimensional_array( $context );
```
(1/3) Supposing we don't do this (don't convert to a multidimensional log) ...

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+	 */
+	public function printable_array( $input, $exclude = array(), $depth = 0 ) {
+		$out   = '';
+		$input = is_object( $input ) ? (array) $input : $input;
```
(2/3) ...won't we anyway get the same result, because we are converting objects to arrays in here...

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+
+				if ( is_array( $val ) || is_object( $val ) ) {
+					$new_depth = $depth + 1;
+					$out .= '[&#010;' . $this->printable_array( $val, $exclude, $new_depth ) . $tabs . ']';
```
(3/3) .. and we rinse and repeat here, in the case of nested arrays/nested objects.  I may be misunderstanding, but wanted to check.

johnregan3 replied:
Great catch.  I think the multidimensional method was a relic from a previous approach I tried. Definitely smooths things out.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+	 *
+	 * @return string
+	 */
+	public function printable_array( $input, $exclude = array(), $depth = 0 ) {
```
I wonder if, rather than add here, this could live in a new utility class (perhaps `Automattic\WooCommerce\Internal\Utilities\Arrays`)?

That would make it easier to use from elsewhere, and would avoid us expanding the public surface area of this class more than needed.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+					hideText = '<?php esc_html_e( 'Hide Details', 'woocommerce' ); ?>';
+
+					// Reset all boxes.
+					$( '.details-box' ).hide();
```
An observation: if I want to compare details from two different rows (perhaps because I'm trying different things while I debug and want to compare them), it's not easy—because if details are already open in one row, clicking View Details in the next row causes the first set of details to be hidden.

This isn't unreasonable, but I wanted to pause and ask what is the preferred/most ideal behavior?

johnregan3 replied:
This is a great question, and your use case makes a lot of sense.  I was concerned about the page possibly becoming hard to manage if you had a lot of rows open -- it would be simpler to have the others automatically close instead of needing to click on each one before you open another.  But, again, what you're saying sounds like a practical scenario.  I'll remove it.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -73,7 +78,7 @@ public function level_dropdown() {
 		);
 
-		$selected_level = isset( $_REQUEST['level'] ) ? $_REQUEST['level'] : '';
+		$selected_level = isset( $_REQUEST['level'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['level'] ) ) : '';
```
Thanks for cleaning this (and other similar items) up!

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+					continue;
+				}
+
+				$tabs = '';
+				for ( $i = 0; $i < $depth; $i++ ) {
+					$tabs .= '  ';
+				}
```
I think we could shorten this and avoid the loop:

```suggestion
				$tabs = str_repeat( '  ', $depth );
```

johnregan3 replied:
Nice recommendation. I have a habit of over-engineering sometimes. :D

johnregan3 replied:
In my updates, I'm moving this declaration outside of the `foreach` loop for efficiency.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+
+				$out .= $tabs . '[' . $key . '] => ';
+
+				if ( is_array( $val ) || is_object( $val ) ) {
+					$new_depth = $depth + 1;
+					$out .= '[&#010;' . $this->printable_array( $val, $exclude, $new_depth ) . $tabs . ']';
+				} else {
+					$out .= "'" . $val . "'";
+				}
```
Possible alternative:

```suggestion
				$out .= is_array( $val )
					? "[\n" . $this->printable_array( $val, $exclude, $depth + 1 ) . $tabs . ']'
					: "'$val'";
```

Thinking here that the `is_object()` check is not needed, as we convert objects to arrays at an earlier stage, and the creation of var `$new_depth` probably doesn't add much readability over `$depth + 1`. We could also apply these thoughts to the existing `if/else` structure, it doesn't need to be changed into a ternary expression.

johnregan3 replied:
I tested this to be sure, and it's possible the array being looped through has objects as values.  Without the `is_object()` check here, things like the `WC_Meta_Data` object cause major problems.

I suppose the `multidimensional_array()` method I removed would have prevented this, as it would walk the entire depth of the array.  However, I think it's much more efficient and versatile to just keep this check here inside `printable_array()`.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+				} else {
+					$out .= "'" . $val . "'";
+				}
+				$out .= '&#013;&#010;';
```
Might not this also work, and arguably be more readable?

```suggestion
				$out .= "\r\n";
```

johnregan3 replied:
This works great!  I believe I was originally using single quotes, which just ended up getting printed as plain text.  I thought it was because I was printing it in a text area.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+				$out .= '&#013;&#010;';
+			}
+		}
+		$out .= '';
```
```suggestion
```

Definitely correct me if I'm wrong, but I don't believe this is needed (since the var is initialized to an empty string at the top of the method, and we only ever append to it, appending a further empty string at the end shouldn't really be needed).

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -392,4 +487,115 @@ protected function prepare_column_headers() {
+				$( '#the-list' ).on( 'click', '.details-button', function( e ) {
+					e.preventDefault();
+
+					var button = $( this ),
+					logId = button.attr( 'data-details-button' ),
+					box = $( "[data-details-box='" + logId + "']" ),
+					defaultText = '<?php esc_html_e( 'View Details', 'woocommerce' ); ?>',
+					hideText = '<?php esc_html_e( 'Hide Details', 'woocommerce' ); ?>';
```
```suggestion
					let $button     = $( this ),
					let logId       = button.attr( 'data-details-button' ),
					let $box        = $( "[data-details-box='" + logId + "']" ),
					let defaultText = '<?php esc_html_e( 'View Details', 'woocommerce' ); ?>',
					let hideText    = '<?php esc_html_e( 'Hide Details', 'woocommerce' ); ?>';
```

The `$` prefix for jQuery objects is a useful convention (and we'd also update any subsequent references).

barryhughes reviewed (changes requested):
Thank you!

I left some notes, but this looks like a nice enhancement.

johnregan3 committed:
Merge branch 'trunk' into 31823-improved-logging-access

johnregan3 committed:
31823 - PR Updates

johnregan3 commented:
Review feedback implemented.

Since this adds the new `src/Admin/Internal/Arrays.php` this might need to be registered somewhere with composer.   I suppose the build process might pick this up, but I'm not sure. 

* Removed the `to_multidimensional_array()` method
* Moved `printable_array()` to an `Arrays` Utility class, then cleaned up the code inside.
* Updated JS to use `$` in var names, and also removed the functionality that closed all rows when a new one was opened.

Thanks for the review, @barryhughes !

barryhughes commented:
> I suppose the build process might pick this up, but I'm not sure.

Yep, it should (I'll check to make sure) and for local dev purposes it may be we need to manually call `composer dump`.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -255,6 +276,83 @@ protected function source_dropdown() {
+	 */
+	public function render_log( $key, $log ) {
+		// This is unserializing "safe" data from the database, not user input.
+		$context = unserialize( $log['context'] );
```
👋🏽 Apologies for the delay. We haven't forgotten about this PR. I'd like to pause and get some further opinions here, as this is a *potential* safety issue (despite the fact it is coming from the database).

I was curious about the using `unserialize()`'s optional `allowed_classes` option as a way of adding safety (we could allow only `stdClass`) but in testing this out it is apparent that that will drop some amount of information, which would defeat the purpose of this change.

A better option may be to switch to using JSON encoding going forward, since this would alleviate safety concerns and it would also be easy to differentiate between JSON and serialized PHP at this stage (we'd avoid visualizing serialized PHP). However, give me a chance to seek some further opinions here before we do more on this :+1:

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -255,6 +276,83 @@ protected function source_dropdown() {
+	 */
+	public function render_log( $key, $log ) {
+		// This is unserializing "safe" data from the database, not user input.
+		$context = unserialize( $log['context'] );
```
👋🏽 Apologies for the delay. We haven't forgotten about this PR. I'd like to pause and get some further opinions here, as this is a *potential* safety issue (despite the fact it is coming from the database).

I was curious about the using `unserialize()`'s optional `allowed_classes` option as a way of adding safety (we could allow only `stdClass`) but in testing this out it is apparent that that will drop some amount of information, which would defeat the purpose of this change.

A better option may be to switch to using JSON encoding going forward, since this would alleviate safety concerns and it would also be easy to differentiate between JSON and serialized PHP at this stage (we'd avoid visualizing serialized PHP). However, give me a chance to seek some further opinions here before we do more on this :+1:

barryhughes replied:
@johnregan3: sorry for the delay.

After talking this out, how about we simply present the context in its serialized form? It will still surface key information (which could for example still be shared with plugin support), is safe, and—if we wished—we might even find a way to pretty-print it.

What are your thoughts?

johnregan3 replied:
@barryhughes We can totally do that. I'd like do this as safely as possible. I'll look into pretty-printing this out.

barryhughes replied:
Excellent, thank you @johnregan3.

barryhughes replied:
As soon as you're ready, please simply re-request a review :+1:

johnregan3 committed:
Update to upstream

johnregan3 committed:
Merge branch 'trunk' into 31823-improved-logging-access

samueljseay commented:
@woocommerce/proton in auditing stale PRs I couldn't decide what to do with this one. It looks like at least the last time it was worked with it was close to ready to merge.

If you could decide next steps for this I'd appreciate it.

barryhughes commented:
Thanks for flagging, @samueljseay! 

It looks like we were waiting on one relatively small change—displaying the context information in 'raw' form, instead of passing it through [unserialize()](https://www.php.net/manual/en/function.unserialize.php)—and otherwise it should be good to go. 

Rather than close it and lose sight of what would otherwise be a nice change, I'll aim to update it over the next couple of weeks (unless of course @johnregan3 you remain interested and have time/space to do so) and move it to completion that way.

johnregan3 commented:
@barryhughes Unfortunately, I don't currently have the bandwidth to work on this right now.  Please proceed as you see fit.  Thanks!

barryhughes commented:
I'll go ahead and close this:

- I don't have bandwidth to take this forward right now.
- There's a lot of good work here, but closing the PR doesn't mean it will be lost: the PR is referenced in the linked issue, and we can revisit the code at any point.
- Once we're ready to pick this back up, thinking that we may want to revise how [`single_row_log`](https://github.com/woocommerce/woocommerce/blob/5c7fd9ad01357c5e539ad51d5890860ca74514cd/plugins/woocommerce/includes/admin/class-wc-admin-log-table-list.php#L298-L317) works: it's true that, besides `source` (which is already surfaced as a table column), the only other significant context vars used in WooCommerce core tend to be `order` and `error`, but we probably shouldn't surface those at the expense of other contextual information that we might want to add in the future, or that third party extensions may add.
